Enable AVX512 Additional 16 SIMD Registers

anthonycanino opened the pull request:
PR enables AVX512 additional 16 SIMD registers for use with the register allocator and codegen emitter. The PR largely consists of the following

0. Extend `xarch` regMask type from `uint32` to `uint64` to make room for additional registers.
1. Add `MM16`-`MM31` to the register definitions (`targetamd.h`).
2. Adjust the `RBM` macros to refer to globally defined variables (float register availability, calleer/callee save register sets etc.) that can be initialized at runtime (so as to add the upper SIMD registers to the definitions if AVX512 is available). Previously, the bank of registers was entirely determined by the architecture, whereas it is now determined by arch and ISA availability).
3. Add a new mode, `JitStressRegs=4` which forces an upper SIMD register to be selected during register allocation.
4. Ensure that an upper SIMD register is selected for a gentree node only if the node can be lowered to an instruction that supports EVEX encoding.
5. Refactor some of the instruction description backend to instead pass around and use an `instrDesc` struct instead of a pure `instruction` to determine when an EVEX prefix is needed. Previously, the instruction itself was enough to determine if SSE/VEX is needed --- now, we require more information about how the instruction is used, e.g., does it require an upper SIMD register.
6. Add logic to emitter to place proper bytes in the EVEX prefix for the upper simd registers.

Of all the points above, (4) requires a little more elaboration. When the LSRA constraints are build up for a node, typically `RBM_NONE` is passed as the constraint mask for an ref interval, which eventually leads to the LSRA selecting the set of available registers for the ref interval type. For nodes whose lowering does not have an available EVEX form (see the commit message), one would like to specify the constraint mask to `RBM_LOWFLOAT` or an equivalent set of registers; however, this came into conflict with the way contained nodes are handling during lowering (and comments even state that the constraint mask feature will not work if contained nodes do not have the same type as the containing node).

To mitigate a lot of rewriting, I created a special `RBM_LOWSIMD` tag in the `regMask` type which functions much like `RBM_NONE` except that during the creation of a new ref position (`LinearScan::newRefPosition`) it will select the low SIMD nodes if the interval is a floating/simd type, otherwise it will default to `RBM_NONE` functionality.

anthonycanino committed:
Change regMask_enum and regMaskTP to unsigned __int64_t on AMD64.

This allows for more registers to be encoded in the register allocator.

anthonycanino committed:
Limit high SIMD reg to compatible intrinsics lsra build.

kunalspathak commented:
Probably we need to figure out first how to fix the regressions showing up in windows/x86:

![image](https://user-images.githubusercontent.com/12488060/207529178-0e9b2090-77e5-4d1a-a746-6320c673d1a9.png)

anthonycanino committed:
Fix for X86 throughput.

tannergooding commented on src/coreclr/jit/emit.h:
```diff
@@ -1121,6 +1121,28 @@ class emitter
+            {
+                case IF_RWR_RRD_RRD:
+                case IF_RWR_RRD_RRD_CNS:
+                case IF_RWR_RRD_RRD_RRD:
```
Is it expected that this one is here and in `idHasReg4`?

tannergooding commented on src/coreclr/jit/target.h:
```diff
@@ -146,13 +150,15 @@ enum _regNumber_enum : unsigned
 {
-    RBM_NONE = 0,
+    RBM_NONE    = 0,
+    RBM_LOWSIMD = 1ULL << 63,
```
I believe we have a macro `UI64(1)` that is preferred here.

It might also be nice if we used `uint64_t` for new code, rather than `unsigned __int64`, but there isn't any consistency there at the moment...

tannergooding commented on src/coreclr/jit/targetamd64.h:
```diff
@@ -78,10 +78,16 @@
 
-  #define RBM_ALLFLOAT            (RBM_XMM0 | RBM_XMM1 | RBM_XMM2 | RBM_XMM3 | RBM_XMM4 | RBM_XMM5 | RBM_XMM6 | RBM_XMM7 | RBM_XMM8 | RBM_XMM9 | RBM_XMM10 | RBM_XMM11 | RBM_XMM12 | RBM_XMM13 | RBM_XMM14 | RBM_XMM15)
+  #define RBM_LOWFLOAT            (RBM_XMM0 | RBM_XMM1 | RBM_XMM2 | RBM_XMM3 | RBM_XMM4 | RBM_XMM5 | RBM_XMM6 | RBM_XMM7 | RBM_XMM8 | RBM_XMM9 | RBM_XMM10 | RBM_XMM11 | RBM_XMM12 | RBM_XMM13 | RBM_XMM14 | RBM_XMM15 )
+  #define RBM_HIGHFLOAT           (RBM_XMM16 | RBM_XMM17 | RBM_XMM18 | RBM_XMM19 | RBM_XMM20 | RBM_XMM21 | RBM_XMM22 | RBM_XMM23 | RBM_XMM24 | RBM_XMM25 | RBM_XMM26 | RBM_XMM27 | RBM_XMM28 | RBM_XMM29 | RBM_XMM30 | RBM_XMM31)
```
I wonder if we could come up with a better name than "HIGHFLOAT"

From the initial `8` SIMD registers, we've had 2 extensions so far and its conceivable a 3rd could happen in the future

anthonycanino replied:
I've used "high" to refer to the 16 AVX512 registers mostly because that is the term reference in the Intel manual, with "next" being used to reference the 8 AVX registers.

I'm in favor of a better naming scheme, but not certain what to choose. We could select `RBM_AVX512FLOAT` or something along those lines...

tannergooding replied:
I don't have a particular preference and I don't think it's "strictly" necessary to update. More a passing thought about how it might be confusing.

anthonycanino commented:
> Probably we need to figure out first how to fix the regressions showing up in windows/x86:
> 
> ![image](https://user-images.githubusercontent.com/12488060/207529178-0e9b2090-77e5-4d1a-a746-6320c673d1a9.png)

I am investigating why we see a throughput hit on x86, it wasn't quite what I thought it was earlier.

BruceForstall commented:
@anthonycanino Looks like you need to fix the formatting.

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=112238&view=ms.vss-build-web.run-extensions-tab) -- no diffs but reasonably significant TP regressions (~0.5%) on x64.

BruceForstall commented:
Do you want to include https://github.com/dotnet/runtime/compare/main...BruceForstall:EnableJitStressHighAvx512Regs?expand=1 in this change?

anthonycanino commented:
> Do you want to include https://github.com/dotnet/runtime/compare/main...BruceForstall:EnableJitStressHighAvx512Regs?expand=1 in this change?

Yes I think that would be a good idea. Do I merge it in this this PR or do we do that seperately?

BruceForstall commented:
> Yes I think that would be a good idea. Do I merge it in this this PR or do we do that seperately?

Include the changes as a new commit. Doesn't need to be a branch merge from mine or anything like that.

anthonycanino committed:
Add upper simd stress test to the AVX512 testing pipeline.

anthonycanino committed:
Add upper 16 SIMD registers to allocator.

Commit includes refactoring code to use `const instrDesc *` instead of `instruction`
so information about when EVEX is needed (due to high SIMD registers) is
available to the emitter.

anthonycanino committed:
Limit high SIMD reg to compatible intrinsics and gentree nodes.

Commit constrains certain hw intrinsics and gentree nodes to use
lower SIMD registers even if upper SIMD registers are available due
to limitations of EVEX encoding for certain instructions.

For example, SSE `Reciprocal` lowers to `rcpps` which does not have an
EVEX encoding form, hence, we cannot allow that hw intrincis node to use
a high SIMD register.

These intrinsics are marked with `HW_Flag_NoEvexSemantics`. Other such
intructions related to masking (typically marked with
`HW_Flag_ReturnsPerElementMask`) also have similar issues (though they
can be replaced with the EVEX k registers and associated masking when
implemented).

In addition, the callee/calleer save registers have also been adjusted
to properly handle the presence and absence of AVX512 upper simd
registers at runtime.

anthonycanino commented:
> Looks like this is likely to conflict with #79478

I've rebased and pushed to test now.

anthonycanino committed:
Formatting.

BruceForstall commented on src/coreclr/jit/hwintrinsic.h:
```diff
@@ -758,6 +764,22 @@ struct HWIntrinsicInfo
+    //
+    static bool HasEvexSemantics(NamedIntrinsic id)
+    {
+        HWIntrinsicFlag flags = lookupFlags(id);
```
Put this inside the `#if`

BruceForstall commented on src/coreclr/jit/target.h:
```diff
@@ -61,7 +61,11 @@ inline bool compUnixX86Abi()
+#define REGMASK_BITS 64
+#define CSE_CONST_SHARED_LOW_BITS 16
+
+#elif defined(TARGET_XARCH)
```
```suggestion
#elif defined(TARGET_X86)
```

BruceForstall commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -11389,33 +11554,33 @@ BYTE* emitter::emitOutputAM(BYTE* dst, instrDesc* id, code_t code, CnsVal* addc)
-    // Explore moving IsWEvexOpcodeExtension() logic inside TakesRexWPrefix(). Not doing so currently
-    // since we cannot differentiate EVEX vs VEX without 'code' until all paths have EVEX support.
-    if (TakesRexWPrefix(ins, size) || (codeEvexMigrationCheck(code) && IsWEvexOpcodeExtension(ins)))
+    // Explore moving IsWEvexOpcodeExtension() logic inside TakesRexWPrefix(). Not doind so currently
+    // since we cannot differentiate EVEX vs VEX without 'code' untill all paths have EVEX support.
```
nit
```suggestion
    // Explore moving IsWEvexOpcodeExtension() logic inside TakesRexWPrefix(). Not doing so currently
    // since we cannot differentiate EVEX vs VEX without 'code' until all paths have EVEX support.
```

BruceForstall commented:
Triggered avx-512 pipeline: https://dev.azure.com/dnceng-public/public/_build/results?buildId=112900&view=results

kunalspathak committed:
Remove unneeded vars

kunalspathak commented on src/coreclr/jit/targetamd64.h:
```diff
@@ -132,15 +142,19 @@
   #define REG_FLT_CALLEE_SAVED_LAST    REG_XMM15
 
-  #define RBM_CALLEE_TRASH        (RBM_INT_CALLEE_TRASH | RBM_FLT_CALLEE_TRASH)
+  #define RBM_CALLEE_TRASH_INIT   (RBM_INT_CALLEE_TRASH | RBM_FLT_CALLEE_TRASH)
```
Isn't there a way to avoid defining all these variables?
Can't you just define the three global variables instead and use them here? See https://github.com/kunalspathak/runtime/commit/08d15493943212d0846118111f54e1719a94763e

anthonycanino replied:
I defined them like that for consistency, but yes I agree for performance your suggestion works much better. I've merged in.

kunalspathak commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -1900,21 +1905,30 @@ int LinearScan::BuildIntrinsic(GenTree* tree)
     }
     assert(tree->gtGetOp2IfPresent() == nullptr);
+
+// TODO-XARCH-AVX512 this is overly constraining register available as NI_System_Math_Abs
```
Can we have this extracted out in a separate method (along with the comment) instead of having this  check at multiple places?

kunalspathak replied:
Also, shouldn't this be under `isEvexCompatible` check instead of just using `RBM_LOWSIMD` for AMD64?

anthonycanino replied:
In this case, its the `NamedIntrinsics` and not the `HWIntrinsics`, e.g., those defined in `namedintrinsiclist.h`.

I suppose I can try and add a similar marking for these but as far as I know there is not a set of flags for these named intrinsics?

kunalspathak commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -2486,33 +2542,41 @@ int LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree, int* pDstCou
-                    srcCount += isRMW ? BuildDelayFreeUses(op3, op1) : BuildOperandUses(op3);
+                    regMaskTP op3RegCandidates = RBM_NONE;
+#if defined(TARGET_AMD64)
+                    if (!isEvexCompatible)
```
This is how I would expect the above code too.

kunalspathak commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -557,6 +557,20 @@ RefPosition* LinearScan::newRefPosition(Interval*    theInterval,
+        if (mask == RBM_LOWSIMD)
+        {
+            // Constrain if we have to for float/simd types
+            if (varTypeIsFloating(theInterval->registerType) || varTypeIsSIMD(theInterval->registerType))
```
This is heavy check to add in `newRefPosition`. Is there a way to move this check (you can have this logic in a helper method or something) to relevant callers which would definitely pass `mask == RBM_LOWSIMD`?

kunalspathak replied:
At the callsite, since you already have the `GenTree*` for the `RefPosition`, you can do something like this:

```c++
regMaskTP mask = RBM_NONE;
if (varTypeIsFloat(node) || varTypeIsSIMD(node)) {
  mask  = lowSIMDRegs();
}

BuildUse(node, mask);

anthonycanino replied:
I believe there was a case where specifying the mask at a `BuildUse` callsite did not cover, something related to a contained `GT_INDIR` node whose register set was not the same as its containing node.

Let me see if I can find the example.

anthonycanino replied:
With this change https://github.com/dotnet/runtime/commit/6cd4d4301957e32af9a318e24428bf33d051bb05 we can see the following example cause some potential issues...

```C#
[MethodImpl(MethodImplOptions.NoInlining)]
static bool Test_Sse41_TestZ_Normal(in Vector128<Int32> x, in Vector128<Int32> y)
{
    return Sse41.TestZ(x, y);
}
```

which lowers to....

```
N001 (  1,  1) [000000] -----------                    t0 =    LCL_VAR   byref  V00 arg0         u:1 (last use) $80
                                                            /--*  t0     byref  
N002 (  7,  5) [000001] ---XG------                    t1 = *  IND       simd16
N003 (  1,  1) [000002] -----------                    t2 =    LCL_VAR   byref  V01 arg1         u:1 (last use) $81
                                                            /--*  t2     byref  
N004 (  7,  5) [000003] -c-XG------                    t3 = *  IND       simd16
                                                            /--*  t1     simd16 
                                                            +--*  t3     simd16 
N005 ( 15, 11) [000004] ---XG------                         *  HWINTRINSIC void   int PTEST <l:$243, c:$242>
```

Since `TestZ` associated lowered instruction `ptest` does not have an EVEX form, we restrict it's operands (which are both float variables that are contained) to the lower SIMD registers. I believe this creates an issue though since if we specify constraints on `t3 = IND` it also propagates to its child node `t2 = LCL_VAR byref ...`. but the nodes have different variable types. Previously, `RBM_NONE` would have been passed down, which would let `newRefPosition` select a set of registers based on ref positions type.

Does this make sense? Essentially, `RBM_LOWSIMD` was a version of `RBM_NONE` to get around this issue.

kunalspathak replied:
Hhm, I thought you were planning to remove it. Did you forget to push?

kunalspathak reviewed (changes requested):
Added some comments for code around LSRA.

anthonycanino committed:
Fix wrong-sized attr for simd mov instruction.

anthonycanino commented:
Will address the comments soon. @dotnet/avx512-contrib can I get an AVX512 pipeline run for correctness check for windows X86?

anthonycanino commented:
/azp run runtime-coreclr jitstress-isas-avx512

dakersnar commented:
/azp run runtime-coreclr jitstress-isas-avx512

anthonycanino committed:
Fix non-AMD64 LSRA stress mask.

anthonycanino commented:
@dotnet/avx512-contrib can I get another avx512 pipeline run?

dakersnar commented:
/azp run runtime-coreclr jitstress-isas-avx512

anthonycanino commented:
Looks like most of the AVX512 test pipeline are passing now --- I will address the comments and then cycle back to 100% pass.

anthonycanino committed:
Update src/coreclr/jit/compiler.h

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Update src/coreclr/jit/compiler.cpp

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Update src/coreclr/jit/gentree.cpp

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Update src/coreclr/jit/hwintrinsic.h

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Update src/coreclr/jit/target.h

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Update src/coreclr/jit/emitxarch.cpp

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino commented on src/coreclr/jit/hwintrinsic.h:
```diff
@@ -172,7 +175,10 @@ enum HWIntrinsicFlag : unsigned int
+    HW_Flag_SupportsContainment = 0x2000,
+
+    // The intrinsic does not have an EVEX compatible form
+    HW_Flag_NoEvexSemantics = 0x4000
```
We don't. I didn't have the `ifdef` done properly, fixed.

anthonycanino committed:
Address PR comments.

anthonycanino commented on src/coreclr/jit/compiler.cpp:
```diff
@@ -3325,6 +3325,30 @@ void Compiler::compInitOptions(JitFlags* jitFlags)
+    rbmProfilerEnterTrash             = RBM_PROFILER_ENTER_TRASH_INIT;
+    rbmProfilerLeaveTrash             = RBM_PROFILER_LEAVE_TRASH_INIT;
+    rbmProfilerTailcallTrash          = RBM_PROFILER_TAILCALL_TRASH_INIT;
+#endif // TARGET_AMD64
```
> So these are global variables now, that are initialized every time the JIT is initialized?
> 
> Can we assume the memory write is atomic? Should they be initialized using InterlockedExchange64? Using Interlocked, we could assert they either have the same value as written, or zero (uninitialized).
> 
> I presume they are not on Compiler because either (1) it was inconvenient to plumb a Compiler* everywhere they are needed, (2) throughput, or (3) you couldn't/didn't want to use the TLS Compiler* (maybe only available in DEBUG?)

Yes, this is why --- passing `compiler` around was both clunky and also incurred its own overhead because it needed to be checked on each access. 

> 
> Could there be a scenario where these change for a single clrjit.dll instance? E.g., what if crossgen2/AOT compilation loads the JIT and compiles both an AVX-512 targeting function and a non-AVX-512 targeting function?

I could use some help here, because honestly I'm not familiar enough with the internals to understand this. As it's done now, I am assuming that these are unchanged and compiling for a single arch/ISA.

> 
> Currently, they are only changed with the JitStress mode. They will presumably be set based on ISA. Does this initialization location have all the appropriate ISA information?

The compiler object has this information as far as I know. Essentially, I would envision replacing with an `compOpportunisticallyDependsOn(InstructionSet_AVX512F_VL)` check.

@tannergooding @kunalspathak are my assumptions here correct? Will this be an issue with `crossgen`?

tannergooding replied:
I think it needs to be filtered on`compOpportunisticallyDependsOn(InstructionSet_AVX512F)`, but I agree that I would expect crossgen to be filtered correctly presuming it.

I do think this needs to be "per compiler instance" and not part of the global state as we can have AltJits or other scenarios where one `compiler` supports an ISA but another does not.

If we could cheapen this further, such as pulling the relevant state into a local where the register allocator needs it, that would presumably be "better".

If the TP regression is "too much" even with that, I could imagine a few ways we could use a template or inlining to ensure the LSRA code still uses a constant and is primarily filtered from a centralized high level check (of course, this comes at the cost of increased deployment size for clrjit.dll).

kunalspathak replied:
> If we could cheapen this further, such as pulling the relevant state into a local where the register allocator needs it

The problem is it is not just LSRA that depends on them, but optimizer and codegen as well. Hence it is little hard to extract that out just at the LSRA level.

kunalspathak replied:
> I presume they are not on Compiler because either (1) it was inconvenient to plumb a Compiler* everywhere they are needed, (2) throughput, or (3) you couldn't/didn't want to use the TLS Compiler* (maybe only available in DEBUG?)

The reason was `#1` , but what you have brought up is a legitimate concern. The only way to avoid it to have them on compiler object and then pass `compiler` instance around. Depending on which area are using these macros, we could do something like, where you have a macro (per file since they are not that many) which takes compiler object, but at all the places where it is needed, you use a version of macro that actually call the one that takes compiler object. That way, at the callsite, you don't have to pass `compiler` object. Its ugly :(

```
#define ABC_WITH_COMPILER(compiler) compiler->isEvex ? x : y
#define ABC ABC_WITH_COMPILER(this)

 = ABC;

```

anthonycanino replied:
If I remember correctly, the throughput overhead of using a compiler instance variable was about 0.50% overhead. It was some of the ugliness of isolating that throughput to X64 that motivated the above approach, i.e. there were _many_ spots where we would have...

```C++
#if defined(TARGET_AMD64)
... = compiler->rbmCalleFltTrash
#else
... = RBM_CALLE_FLT_TRASH
#endif
```

I believe we can use the above approach to mitigate some of this but it will still be a little ugly...

```C++
#if defined(TARGET_AMD64)
#define RBM_CALLE_FLT_TRASH_USE compiler->rbmCalleFltTrash
#else
#define RBM_CALLE_FLT_TRASH_USE RBM_CALLE_FLT_TRASH
#endif
```

anthonycanino replied:
@kunalspathak @tannergooding do you have any more thoughts on how to adjust the PR for this?

kunalspathak replied:
Did you try the approach in https://github.com/dotnet/runtime/pull/79544#discussion_r1063850725 and https://github.com/dotnet/runtime/pull/79544#discussion_r1063883720?

anthonycanino replied:
Ok, let me try this approach https://github.com/dotnet/runtime/pull/79544#discussion_r1063883720 then and see how the numbers look.

anthonycanino replied:
@kunalspathak I have a draft of this up at https://github.com/dotnet/runtime/pull/80559/commits/9c386bc654bedeced1c05b66e9a0bbb4a581b70e . Currently waiting for test data / throughput.

To be honest, I do not like it, it seems really messy (even if we move the macro definitions to the call sites). It also seems the alternatives aren't great either.

anthonycanino replied:
FYI, with the change https://github.com/dotnet/runtime/commit/9c386bc654bedeced1c05b66e9a0bbb4a581b70e ...

![image](https://user-images.githubusercontent.com/91913526/212180309-6965b96b-d6b2-421a-a913-719c8a31a742.png)

It seems like most approaches have been pretty similar with throughput impact.

anthonycanino committed:
Allow `emitinl.h` access to the `rbm` variables.

kunalspathak commented on src/coreclr/jit/targetamd64.h:
```diff
@@ -220,8 +234,10 @@
+  #define CNT_CALLEE_SAVED_FLOAT        (10)
+  #define CNT_CALLEE_TRASH_FLOAT_INIT   (6)
+  /* NOTE: Sync with variable name defined in compiler.h */
+  #define CNT_CALLEE_TRASH_FLOAT   rbmCntCalleeTrashFloat
```
This and the one on line 228 can be removed outside of `#ifdef-else` ?

kunalspathak commented on src/coreclr/jit/targetamd64.h:
```diff
@@ -132,15 +142,18 @@
-  #define RBM_FLT_CALLEE_TRASH    (RBM_XMM0|RBM_XMM1|RBM_XMM2|RBM_XMM3|RBM_XMM4|RBM_XMM5)
+  /* NOTE: Sync with variable name defined in compiler.h */
+  #define RBM_FLT_CALLEE_TRASH_INIT (RBM_XMM0|RBM_XMM1|RBM_XMM2|RBM_XMM3|RBM_XMM4|RBM_XMM5)
+  #define RBM_FLT_CALLEE_TRASH    rbmFltCalleeTrash
```
Same here - this can be outside of `#ifdef-else`.

anthonycanino committed:
Replace RBM_LOWSIMD with `BuildEvexIncompatibleMask`.

anthonycanino committed:
Move AVX512 dependent `targetamd.h` vars into compiler object.

kunalspathak commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -1885,21 +1890,30 @@ int LinearScan::BuildIntrinsic(GenTree* tree)
+// TODO-XARCH-AVX512 this is overly constraining register available as NI_System_Math_Abs
+// can be lowered to EVEX compatible instruction (the rest cannot)
+#if defined(TARGET_AMD64)
+    regMaskTP opRegMask = RBM_LOWSIMD;
```
possibly have a macro like `#define DEFAULT_CANDIDATE_MASK` and set it to `RBM_LOWSIMD` vs. `RBM_NONE`?

anthonycanino replied:
I'm not sure this will apply after the changes, since now we isolate to the `BuildEvexIncompatibleMask` function. Can make adjustments after I push if need be, but I think it'll be ok how it is.

anthonycanino committed:
Fixing some edge cases for `targetamd.h` variables.

anthonycanino committed:
Limit high SIMD reg to compatible intrinsics lsra build.

anthonycanino committed:
Fix a merge/rebase bug.

anthonycanino commented:
There were some superpmi diffs which I have fixed (I accidently left some intrinsics that a commit on `main` removed). There is a couple failures on the JIT ISA stress test that I will address.

In the mean time, can I get some review on the design changes, e.g., `BuildEvexIncompatibleMask` and the `RBM_ALLFLOAT_USE` macro/compiler variables changes.

kunalspathak commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -8929,8 +8975,18 @@ void LinearScan::dumpLsraStatsSummary(FILE* file)
-void dumpRegMask(regMaskTP regs)
+// Please see the comment for these instance variables in `compiler.h`
+#if defined(TARGET_AMD64)
+#define RBM_ALLFLOAT_USE (compiler->rbmAllFloat)
```
why can't you use the macros previously defined and undefed few lines back?

anthonycanino replied:
I should leave a comment here: this `compiler` must match the parameter in `dumpRegMask`. I thought it more error prone to assume the name passed in here is the same one at the class level for the file.

kunalspathak commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -2701,4 +2739,25 @@ void LinearScan::SetContainsAVXFlags(unsigned sizeOfSIMDVector /* = 0*/)
     }
 }
 
+inline regMaskTP LinearScan::BuildEvexIncompatibleMask(GenTree* tree)
```
can we have `assert(!tree->isEvexCompatibleHWIntrinsic())`?

anthonycanino replied:
We can't, because it's not always an intrinsic. For example, `GT_CNS_VEC` cannot use EVEX form as it is currently lowered, but it is not a HWIntrinsic.

kunalspathak reviewed (changes requested):
> can I get some review on the design changes, e.g., BuildEvexIncompatibleMask and the RBM_ALLFLOAT_USE macro/compiler variables changes.

The changes looks good to me. @BruceForstall , in case you want to check it out too.

What are your plans to address the TP regressions?

![image](https://user-images.githubusercontent.com/12488060/213381573-51c1de4a-aa8a-42cf-9aaf-1997238368ff.png)

anthonycanino commented:
> > can I get some review on the design changes, e.g., BuildEvexIncompatibleMask and the RBM_ALLFLOAT_USE macro/compiler variables changes.
> 
> The changes looks good to me. @BruceForstall , in case you want to check it out too.
> 
> What are your plans to address the TP regressions?
> 
> ![image](https://user-images.githubusercontent.com/12488060/213381573-51c1de4a-aa8a-42cf-9aaf-1997238368ff.png)

I think a lot of the throughput impact came from https://github.com/dotnet/runtime/pull/79544/commits/9da5b6d2ca3289c080f6af8f13eb73de6035af28 where we effectively doubled the size of the `regMaskTP` data structure to accommodate more registers.

From discussions, it looks like mitigating this throughput impact would require some significant engineering effort (we would have to make register sets pluggable) etc. Unless there are other, simpler ideas here that I am missing.

@BruceForstall do you have thoughts or suggestions?

kunalspathak commented:
> I think a lot of the throughput impact came from https://github.com/dotnet/runtime/commit/9da5b6d2ca3289c080f6af8f13eb73de6035af28 where we effectively doubled the size of the regMaskTP data structure to accommodate more registers.

Let me double check what methods are impacted the most and we can figure out a way. Most of the `regMaskTP` does is find the first bit set or `regNumber` from the mask using `log2()` so that should not be impacted. Probably, the ones that iterate over the bits might be impacting and we could add logic there to just iterate over lower 32-bits if EVEX is not supported because upper 32 bits will always be zero.

tannergooding commented:
> Probably, the ones that iterate over the bits might be impacting and we could add logic there to just iterate over lower 32-bits if EVEX is not supported because upper 32 bits will always be zero.

What functions are iterating over bits?

I'd expect us to use optimized Lzcnt, Tzcnt, and Popcnt functions for many such things. Lzcnt and Tzcnt should be 1 instruction on x86/x64 (bsf/bsr) + a branch handling zero, if zero is possible. Likewise `Popcnt` can be 10-11 instructions (not a naive loop) and can lightup to be 1 instruction with a cached CPUID check.

kunalspathak commented:
> What functions are iterating over bits?

I have to find that out. Usually when we iterate, we have pattern that looks something like this which shouldn't matter what the size of `regMaskTP`.

```c++
    while (targetCandidates != RBM_NONE)
    {
        regMaskTP targetRegMask = genFindLowestBit(targetCandidates);
        ...
    }

```

tannergooding commented:
Yeah, that shouldn't be super impactful for 64-bit. I could see it being somewhat impactful to 32-bit.

That being said, it looks like a common pattern is basically "get lowest bit from y", "toggle off lowest bit in y", "use log2 of mask to get regNum". So we basically have a sequence of:
```
; regMaskTP targetRegMask = genFindLowestBit(targetCandidates)
mov
neg
and

; targetCandidates &= ~targetRegMask;
mov
not
and

; regNumber targetReg = genRegNumFromMask(targetRegMask);
bsf
```

If you rely on the fact that we know `targetRegMask` is a set bit in `targetCandidates` and we know we want to toggle it off, we can change it to:
```
; regMaskTP targetRegMask = genFindLowestBit(targetCandidates)
mov
neg
and

; targetCandidates ^= targetRegMask;
xor

; regNumber targetReg = genRegNumFromMask(targetRegMask);
bsf
```

However, even more so if we invert this to be `regNumber genFindLowestRegNum(regMaskTP mask)` we can have something like
```
; regNumber targetReg = genRegNumFromMask(targetCandidates);
bsf

; targetCandidates  = BitTestAndComplement(targetCandidates, targetReg);
btc
```

We keep the existing `bsf` which is ~3 cycles and we utilize the super efficient `btc` which is 1 cycle and complements the bit at specified index
In the case we don't know that the bit is set, we can use `BitTestAndReset` which achieves the same thing and is likewise `1 cycle`.

tannergooding commented:
Also worth noting that `genLog2` is using `BitPosition` which is using `__builtin_ffsll` on Unix. It is much more efficient to use `__builtin_ctzll`

We likewise don't seem to have helpers for `BTS`, `BTC`, or `BTR` (set, complement, and reset selected bit, respectively). MSVC, Clang, and GCC all light these up selectively in some cases (such as when the `selected bit` is known to come from `BSF` or `BSR`). However, explicit helpers can ensure this behavior and help when the compiler analysis can't track that the given bit is a valid index.

kunalspathak commented:
> Also worth noting that `genLog2` is using `BitPosition` which is using `__builtin_ffsll` on Unix. It is much more efficient to use `__builtin_ctzll`
> 
> We likewise don't seem to have helpers for `BTS`, `BTC`, or `BTR` (set, complement, and reset selected bit, respectively). MSVC, Clang, and GCC all light these up selectively in some cases (such as when the `selected bit` is known to come from `BSF` or `BSR`). However, explicit helpers can ensure this behavior and help when the compiler analysis can't track that the given bit is a valid index.

Yes, that's what I am seeing. Let me try experimenting something and see how much gain we can get.

kunalspathak commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -21,6 +21,13 @@ XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+#if defined(TARGET_AMD64)
+#define RBM_ALLFLOAT_USE (compiler->rbmAllFloat)
+#define RBM_FLT_CALLEE_TRASH_USE (compiler->rbmFltCalleeTrash)
+#define CNT_CALLEE_TRASH_FLOAT_USE (compiler->cntCalleeTrashFloat)
```
`CNT_CALLEE_TRASH_FLOAT` is not used in most of the places where this macro is defined. I would recommend to define it only where it is used like `optimizer.cpp`.

kunalspathak commented:
I tried to repro TP regression for benchmarks on windows/x64 and they are significant, specially for `MinOpts`. 


<details>
<summary>Overall (<span style="color:red">+2.40%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+2.40%</span>|


</details>

<details>
<summary>MinOpts (<span style="color:red">+9.02%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+9.02%</span>|


</details>

<details>
<summary>FullOpts (<span style="color:red">+2.35%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+2.35%</span>|


</details>

<details>
<summary>Details</summary>

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.windows.x64.checked.mch|53,679,612,879|54,969,533,478|<span style="color:red">+2.40%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.windows.x64.checked.mch|441,545,663|481,383,997|<span style="color:red">+9.02%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|benchmarks.run.windows.x64.checked.mch|53,238,067,216|54,488,149,481|<span style="color:red">+2.35%</span>|


</details>


Here are some of the methods that are contributing:

```
private: class RefPosition * __ptr64 __cdecl LinearScan::newRefPositionRaw(unsigned int,struct GenTree * __ptr64,enum RefType) __ptr64                      
private: class RefPosition * __ptr64 __cdecl LinearScan::newRefPosition(class Interval * __ptr64,unsigned int,enum RefType,struct GenTree * __ptr64,unsigned __int64,unsigned int) __ptr64
private: void __cdecl LinearScan::associateRefPosWithInterval(class RefPosition * __ptr64) __ptr64                                                                     
private: class RefPosition * __ptr64 __cdecl LinearScan::newRefPosition(enum _regNumber_enum,unsigned int,enum RefType,struct GenTree * __ptr64,unsigned __int64) __ptr64                                                                                 
private: class RefPosition * __ptr64 __cdecl LinearScan::BuildDef(struct GenTree * __ptr64,unsigned __int64,int) __ptr64   
```

I also looked in `newRefPositionRaw` and seems like previously it was getting inlined more often and with this PR, it is not.

Here is the callstack in base vs. diff:

![image](https://user-images.githubusercontent.com/12488060/213626525-931d838e-32e4-46b8-81bb-443d0a83ac3b.png)

![image](https://user-images.githubusercontent.com/12488060/213626555-24d48aff-f893-4743-a1bb-97375995ac12.png)

@anthonycanino - can you see what `tpdiff` shows for you locally?

anthonycanino commented:
> @anthonycanino - can you see what `tpdiff` shows for you locally?

I ran locally, this is what I get...

Warning: Different compilers used for base and diff JITs. Results may be misleading.
Base JIT's compiler: MSVC 193431933
Diff JIT's compiler: MSVC 193331629

<details>
<summary>Overall (<span style="color:red">+0.88%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:green">-0.31%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.22%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+1.53%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.43%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.26%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.58%</span>|


</details>

<details>
<summary>MinOpts (<span style="color:red">+2.24%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.47%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+1.75%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+2.38%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.64%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+1.58%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:green">-0.37%</span>|


</details>

<details>
<summary>FullOpts (<span style="color:red">+0.52%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:green">-0.42%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.20%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.84%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.43%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.25%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.60%</span>|


</details>

<details>
<summary>Details</summary>

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|122,112,357,067|121,739,041,645|<span style="color:green">-0.31%</span>|
|benchmarks.run.windows.x64.checked.mch|35,857,662,517|35,935,616,729|<span style="color:red">+0.22%</span>|
|coreclr_tests.run.windows.x64.checked.mch|817,127,484,937|829,652,803,938|<span style="color:red">+1.53%</span>|
|libraries.crossgen2.windows.x64.checked.mch|129,243,644,472|129,796,388,884|<span style="color:red">+0.43%</span>|
|libraries.pmi.windows.x64.checked.mch|235,404,739,678|236,018,090,946|<span style="color:red">+0.26%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|504,105,025,340|507,022,605,960|<span style="color:red">+0.58%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|15,231,627,619|15,303,658,848|<span style="color:red">+0.47%</span>|
|benchmarks.run.windows.x64.checked.mch|473,937,212|482,224,103|<span style="color:red">+1.75%</span>|
|coreclr_tests.run.windows.x64.checked.mch|367,582,440,292|376,349,064,057|<span style="color:red">+2.38%</span>|
|libraries.crossgen2.windows.x64.checked.mch|1,742,708|1,753,844|<span style="color:red">+0.64%</span>|
|libraries.pmi.windows.x64.checked.mch|1,436,504,168|1,459,168,629|<span style="color:red">+1.58%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|9,623,139,135|9,587,249,160|<span style="color:green">-0.37%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|106,880,729,448|106,435,382,797|<span style="color:green">-0.42%</span>|
|benchmarks.run.windows.x64.checked.mch|35,383,725,305|35,453,392,626|<span style="color:red">+0.20%</span>|
|coreclr_tests.run.windows.x64.checked.mch|449,545,044,645|453,303,739,881|<span style="color:red">+0.84%</span>|
|libraries.crossgen2.windows.x64.checked.mch|129,241,901,764|129,794,635,040|<span style="color:red">+0.43%</span>|
|libraries.pmi.windows.x64.checked.mch|233,968,235,510|234,558,922,317|<span style="color:red">+0.25%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|494,481,886,205|497,435,356,800|<span style="color:red">+0.60%</span>|


</details>

anthonycanino commented:
@kunalspathak I think this is correct...


<details>
<summary>Overall (<span style="color:red">+3.47%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+3.22%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+2.74%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+3.99%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+3.68%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+2.88%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+2.96%</span>|


</details>

<details>
<summary>MinOpts (<span style="color:red">+5.66%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+7.29%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+9.02%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+5.53%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+6.38%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+8.60%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+7.78%</span>|


</details>

<details>
<summary>FullOpts (<span style="color:red">+2.88%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+2.65%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+2.66%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+2.74%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+3.68%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+2.84%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+2.87%</span>|


</details>

<details>
<summary>Details</summary>

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|117,946,571,536|121,739,137,285|<span style="color:red">+3.22%</span>|
|benchmarks.run.windows.x64.checked.mch|34,977,691,827|35,935,434,733|<span style="color:red">+2.74%</span>|
|coreclr_tests.run.windows.x64.checked.mch|797,827,556,315|829,652,030,786|<span style="color:red">+3.99%</span>|
|libraries.crossgen2.windows.x64.checked.mch|125,191,945,405|129,796,552,061|<span style="color:red">+3.68%</span>|
|libraries.pmi.windows.x64.checked.mch|229,420,486,736|236,017,919,645|<span style="color:red">+2.88%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|492,440,980,002|507,024,440,844|<span style="color:red">+2.96%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|14,263,303,643|15,303,658,761|<span style="color:red">+7.29%</span>|
|benchmarks.run.windows.x64.checked.mch|442,315,219|482,223,980|<span style="color:red">+9.02%</span>|
|coreclr_tests.run.windows.x64.checked.mch|356,628,667,746|376,348,991,286|<span style="color:red">+5.53%</span>|
|libraries.crossgen2.windows.x64.checked.mch|1,648,632|1,753,842|<span style="color:red">+6.38%</span>|
|libraries.pmi.windows.x64.checked.mch|1,343,642,030|1,459,168,368|<span style="color:red">+8.60%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|8,895,125,188|9,587,255,330|<span style="color:red">+7.78%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|103,683,267,893|106,435,478,524|<span style="color:red">+2.65%</span>|
|benchmarks.run.windows.x64.checked.mch|34,535,376,608|35,453,210,753|<span style="color:red">+2.66%</span>|
|coreclr_tests.run.windows.x64.checked.mch|441,198,888,569|453,303,039,500|<span style="color:red">+2.74%</span>|
|libraries.crossgen2.windows.x64.checked.mch|125,190,296,773|129,794,798,219|<span style="color:red">+3.68%</span>|
|libraries.pmi.windows.x64.checked.mch|228,076,844,706|234,558,751,277|<span style="color:red">+2.84%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|483,545,854,814|497,437,185,514|<span style="color:red">+2.87%</span>|


</details>

kunalspathak commented:
This looks similar to what I saw. Are you doing this on windows or Linux? Let me know if you need any help in figuring it out. Perhaps comment/revert certain commits to narrow down the source.

<html>
<body>
<!--StartFragment-->

benchmarks.run.windows.x64.checked.mch | +9.02%
-- | --


<!--EndFragment-->
</body>
</html>

BruceForstall commented on src/coreclr/jit/targetamd64.h:
```diff
@@ -203,15 +217,18 @@
 #endif
 
-  #define REG_VAR_ORDER_FLT      REG_XMM0,REG_XMM1,REG_XMM2,REG_XMM3,REG_XMM4,REG_XMM5,REG_XMM6,REG_XMM7,REG_XMM8,REG_XMM9,REG_XMM10,REG_XMM11,REG_XMM12,REG_XMM13,REG_XMM14,REG_XMM15
+  //#define REG_VAR_ORDER_FLT      REG_XMM0,REG_XMM1,REG_XMM2,REG_XMM3,REG_XMM4,REG_XMM5,REG_XMM6,REG_XMM7,REG_XMM8,REG_XMM9,REG_XMM10,REG_XMM11,REG_XMM12,REG_XMM13,REG_XMM14,REG_XMM15
```
Remove commented out code before merge

BruceForstall commented on src/coreclr/jit/targetamd64.h:
```diff
@@ -203,15 +217,18 @@
 
-  #define REG_VAR_ORDER_FLT      REG_XMM0,REG_XMM1,REG_XMM2,REG_XMM3,REG_XMM4,REG_XMM5,REG_XMM6,REG_XMM7,REG_XMM8,REG_XMM9,REG_XMM10,REG_XMM11,REG_XMM12,REG_XMM13,REG_XMM14,REG_XMM15
+  //#define REG_VAR_ORDER_FLT      REG_XMM0,REG_XMM1,REG_XMM2,REG_XMM3,REG_XMM4,REG_XMM5,REG_XMM6,REG_XMM7,REG_XMM8,REG_XMM9,REG_XMM10,REG_XMM11,REG_XMM12,REG_XMM13,REG_XMM14,REG_XMM15
+  #define REG_VAR_ORDER_FLT      REG_XMM0,REG_XMM1,REG_XMM2,REG_XMM3,REG_XMM4,REG_XMM5,REG_XMM6,REG_XMM7,REG_XMM8,REG_XMM9,REG_XMM10,REG_XMM11,REG_XMM12,REG_XMM13,REG_XMM14,REG_XMM15,REG_XMM16,REG_XMM17,REG_XMM18,REG_XMM19,REG_XMM20,REG_XMM22,REG_XMM23,REG_XMM24,REG_XMM25,REG_XMM26,REG_XMM27,REG_XMM28,REG_XMM29,REG_XMM30,REG_XMM31
```
nit: Consider formatting so it all fits in a 120-character line. Maybe 8 regs per line?

BruceForstall commented on src/coreclr/jit/codegencommon.cpp:
```diff
@@ -9412,3 +9429,7 @@ bool CodeGen::genCanOmitNormalizationForBswap16(GenTree* tree)
     return (cast->gtCastType == TYP_USHORT) || (cast->gtCastType == TYP_SHORT);
 }
+
+#undef RBM_ALLFLOAT_USE
+#undef RBM_FLT_CALLEE_TRASH_USE
+#undef CNT_CALLEE_TRASH_FLOAT_USE
```
You don't need to worry about `#undef` at the end of a translation unit. I suppose it's ok, but unnecessary?

BruceForstall commented on src/coreclr/jit/lsra.cpp:
```diff
@@ -671,6 +706,17 @@ LinearScan::LinearScan(Compiler* theCompiler)
 #endif // TARGET_AMD64 || TARGET_ARM64
 
+#if defined(TARGET_AMD64)
+    // TODO-XARCH-AVX512 switch this to canUseEvexEncoding() once we independetly
```
nit
```suggestion
    // TODO-XARCH-AVX512 switch this to canUseEvexEncoding() once we independently
```
```suggestion
    // TODO-XARCH-AVX512 switch this to canUseEvexEncoding() once we independetly
```

BruceForstall reviewed (commented):
Added a few minor comments.

The `RBM_*_USE` defines seem reasonable.

Perhaps another alternative is to define:
```
#if defined(TARGET_AMD64)
#define RBM_ALLFLOAT_USE (this->rbmAllFloat)
#define RBM_FLT_CALLEE_TRASH_USE (this->rbmFltCalleeTrash)
#define CNT_CALLEE_TRASH_FLOAT_USE (this->cntCalleeTrashFloat)
#endif
```

always, and just create a class-local copy of rbmAllFloat, etc., in each class that references it (CodeGen, emitter, LinearScan) -- each of these constructors would copy the value from Compiler (that is set once during initialization). That might reduce the amount of `#define`/`#undef` magic, and referencing, say, rbmAllFloat would go from `this->compiler->rbmAllFloat` to just `this->rbmAllFloat` in some cases.

Maybe there are some "smaller" classes (compared to CodeGen, emitter, LinearScan, which are all single instance) where it wouldn't be worthwhile to make a copy, in which case the current method would be still used.

anthonycanino committed:
Update src/coreclr/jit/compiler.h

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Update src/coreclr/jit/lsra.cpp

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Update src/coreclr/jit/compiler.h

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

anthonycanino committed:
Fix nits.

anthonycanino committed:
Merge branch 'avx512-upper-regs' of https://github.com/anthonycanino/runtime into avx512-upper-regs

anthonycanino commented:
FYI, after disabling PGO and comparing with base/diff, we see the following throughput changes...


<details>
<summary>Overall (<span style="color:red">+0.48%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.67%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.62%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.36%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.72%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.65%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.48%</span>|


</details>

<details>
<summary>MinOpts (<span style="color:red">+0.40%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.66%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.55%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.38%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+1.14%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.62%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.35%</span>|


</details>

<details>
<summary>FullOpts (<span style="color:red">+0.50%</span>)</summary>

|Collection|PDIFF|
|---|--:|
|aspnet.run.windows.x64.checked.mch|<span style="color:red">+0.67%</span>|
|benchmarks.run.windows.x64.checked.mch|<span style="color:red">+0.62%</span>|
|coreclr_tests.run.windows.x64.checked.mch|<span style="color:red">+0.33%</span>|
|libraries.crossgen2.windows.x64.checked.mch|<span style="color:red">+0.72%</span>|
|libraries.pmi.windows.x64.checked.mch|<span style="color:red">+0.65%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|<span style="color:red">+0.49%</span>|


</details>

<details>
<summary>Details</summary>

All contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|122,093,536,277|122,913,937,175|<span style="color:red">+0.67%</span>|
|benchmarks.run.windows.x64.checked.mch|35,841,181,058|36,063,565,868|<span style="color:red">+0.62%</span>|
|coreclr_tests.run.windows.x64.checked.mch|816,959,925,687|819,862,183,488|<span style="color:red">+0.36%</span>|
|libraries.crossgen2.windows.x64.checked.mch|129,220,305,672|130,145,275,216|<span style="color:red">+0.72%</span>|
|libraries.pmi.windows.x64.checked.mch|235,372,957,710|236,893,696,628|<span style="color:red">+0.65%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|504,076,931,241|506,511,326,375|<span style="color:red">+0.48%</span>|

MinOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|15,228,578,840|15,329,437,421|<span style="color:red">+0.66%</span>|
|benchmarks.run.windows.x64.checked.mch|473,829,608|476,421,325|<span style="color:red">+0.55%</span>|
|coreclr_tests.run.windows.x64.checked.mch|367,540,737,175|368,953,049,912|<span style="color:red">+0.38%</span>|
|libraries.crossgen2.windows.x64.checked.mch|1,742,113|1,762,012|<span style="color:red">+1.14%</span>|
|libraries.pmi.windows.x64.checked.mch|1,436,142,118|1,445,036,143|<span style="color:red">+0.62%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|9,621,502,879|9,654,929,734|<span style="color:red">+0.35%</span>|

FullOpts contexts:

|Collection|Base # instructions|Diff # instructions|PDIFF|
|---|--:|--:|--:|
|aspnet.run.windows.x64.checked.mch|106,864,957,437|107,584,499,754|<span style="color:red">+0.67%</span>|
|benchmarks.run.windows.x64.checked.mch|35,367,351,450|35,587,144,543|<span style="color:red">+0.62%</span>|
|coreclr_tests.run.windows.x64.checked.mch|449,419,188,512|450,909,133,576|<span style="color:red">+0.33%</span>|
|libraries.crossgen2.windows.x64.checked.mch|129,218,563,559|130,143,513,204|<span style="color:red">+0.72%</span>|
|libraries.pmi.windows.x64.checked.mch|233,936,815,592|235,448,660,485|<span style="color:red">+0.65%</span>|
|libraries_tests.pmi.windows.x64.checked.mch|494,455,428,362|496,856,396,641|<span style="color:red">+0.49%</span>|


</details>

anthonycanino committed:
Trying VM changes.

anthonycanino committed:
VM hack.

anthonycanino committed:
VM hack.

kunalspathak commented:
I just realized that increasing the size of `regMaskTP` would also increase the size of each `RefPosition` and `Interval` by 4 bytes.

anthonycanino commented:
> I just realized that increasing the size of `regMaskTP` would also increase the size of each `RefPosition` and `Interval` by 4 bytes each.

I believe we knew that there would be some memory impact and measured it a while back. I can dig up the numbers.

anthonycanino committed:
Revert "VM hack."

This reverts commit 91cf3db9115e94ca1d759045fe4da6a496228cfc.

anthonycanino committed:
Adjust ACTUAL_REG_COUNT based on availability of AVX512.

kunalspathak commented:
Latest TP numbers:

![image](https://user-images.githubusercontent.com/12488060/217094799-48b21489-b04f-4943-a0a3-10c3537ed0a1.png)

anthonycanino commented:
@dotnet/avx512-contrib can I get a rerun of the avx512 stress pipelines?

tannergooding commented:
Requeued the failing jobs.

anthonycanino commented:
> Latest TP numbers:
> 
> ![image](https://user-images.githubusercontent.com/12488060/217094799-48b21489-b04f-4943-a0a3-10c3537ed0a1.png)

What are our thoughts on this? Are we in acceptable range?

tannergooding commented:
-- Just worth noting that between November and now we've gotten multiply TP improvements in, including some that were `-0.4%` range, so we won't be regressing anything as compared to .NET 7

There are likely more TP improvements to be had for this code of course, but it's potentially something we can iterate on separately.

kunalspathak commented:
> but it's potentially something we can iterate on separately.

I agree.

anthonycanino commented:
So are we ok with the changes now?

BruceForstall commented:
Can you remind me why there is any TP impact on x86?

anthonycanino commented:
> Can you remind me why there is any TP impact on x86?
> 
> [Edit] I suppose that could be due to the `instruction ins` => `const instrDesc* id` parameter change, for example.

Yes, this is what I think.

BruceForstall committed:
Use inline accessor functions instead of macros

Convert from macros to accessor functions for
RBM_ALLFLOAT, RBM_FLT_CALLEE_TRASH, CNT_CALLEE_TRASH_FLOAT.
Convert LSRA use of ACTUAL_REG_COUNT to AVAILABLE_REG_COUNT,
and create an accessor for that value for AMD64 as well.

BruceForstall commented:
The macros defining the AVX512-specific masks and counts were bothering me, so I wrote https://github.com/dotnet/runtime/pull/81818, to implement them as accessor functions in the various classes that need them. I think it looks better this way. Comments?

kunalspathak commented on src/coreclr/jit/lsrabuild.cpp:
```diff
@@ -1860,8 +1868,9 @@ void LinearScan::buildRefPositionsForNode(GenTree* tree, LsraLocation currentLoc
-const unsigned         lsraRegOrderSize    = ArrLen(lsraRegOrder);
+static const regNumber lsraRegOrder[]   = {REG_VAR_ORDER};
+const unsigned         lsraRegOrderSize = ArrLen(lsraRegOrder);
+// TODO-XARCH-AVX512 we might want to move this to be configured with the rbm variables too
```
Should this happen in this PR?

anthonycanino replied:
Just noting will do in a follow up PR

kunalspathak commented on src/coreclr/jit/lsraxarch.cpp:
```diff
@@ -2701,4 +2739,25 @@ void LinearScan::SetContainsAVXFlags(unsigned sizeOfSIMDVector /* = 0*/)
+        return RBM_NONE;
+    }
+
+    if (tree->isContained() &&
```
Worthwhile to add a comment for this case.

kunalspathak reviewed (commented):
Register allocator portion looks good to me modulo the changes suggested by @BruceForstall  in #81818.

anthonycanino commented:
> The macros defining the AVX512-specific masks and counts were bothering me, so I wrote #81818, to implement them as accessor functions in the various classes that need them. I think it looks better this way. Comments?
> 
> [Edit] I created a PR to this overall change if you decide to take it: [anthonycanino#6](https://github.com/anthonycanino/runtime/pull/6)

Yes, I will pull in.

anthonycanino committed:
Merge pull request #6 from BruceForstall/anthonycanino_avx512-upper-regs-with-reg-accessors

Use inline accessor functions instead of macros

anthonycanino committed:
Clearifying comments.

BruceForstall commented:
[No diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=164641&view=ms.vss-build-web.run-extensions-tab), TP regressions as before.

Failures looks like infra or unrelated.

kunalspathak reviewed (approved):
LGTM. Thanks for your patience and efforts!

anthonycanino commented:
Thanks all!
